examples/extra_element_integer/clustering_algorithm.cpp

```diff
@@ -50,53 +50,62 @@ void AddVariablesToSystem(libMesh::EquationSystems &equation_systems,
+                              const libMesh::Elem *elem,
                               const unsigned int variable_num) {
 
   DofMap &dof_map = system.get_dof_map();
   std::vector<dof_id_type> dof_indices;
-  dof_map.dof_indices(&elem, dof_indices, variable_num);
+  dof_map.dof_indices(elem, dof_indices, variable_num);
 
   std::vector<double> solution_value(1);
   system.solution->get(dof_indices, solution_value);
 
   return static_cast<double>(solution_value[0]);
 }
 
+bool BelongToCluster(libMesh::Elem *elem, libMesh::Elem *neighbor_elem,
+                     int variable_num, libMesh::LinearImplicitSystem &system) {
+
+  double element_solution = GetElementDataFromMesh(system, elem, variable_num);
+  double neighbor_solution =
+      GetElementDataFromMesh(system, neighbor_elem, variable_num);
+  if (static_cast<int>(element_solution) ==
+      static_cast<int>(neighbor_solution)) {
+
+    return true;
+  } else {
+    return false;
+  }
+}
+
 const unsigned int FindCluster(libMesh::Mesh &mesh,
-                               LinearImplicitSystem &system,
+                               libMesh::LinearImplicitSystem &system,
                                const std::string &variable_name) {
-
   const unsigned int variable_num = system.variable_number(variable_name);
-
-  const unsigned int index = mesh.add_elem_integer(variable_name);
+  int not_visited = -1;
+  const unsigned int index = mesh.add_elem_integer(variable_name, not_visited);
+  std::stack<libMesh::Elem *> neighbor_stack;
 
   for (const auto &elem : mesh.element_ptr_range()) {
-
-    int element_solution =
-        static_cast<int>(GetElementDataFromMesh(system, *elem, variable_num));
-
-    bool belong_to_a_cluster = false;
-    for (unsigned int side = 0; side < elem->n_sides(); ++side) {
-      const Elem *neighbor = elem->neighbor_ptr(side);
-      if (neighbor) {
-
-        int neighbor_element_solution = static_cast<int>(
-            GetElementDataFromMesh(system, *neighbor, variable_num));
-
-        if (element_solution == neighbor_element_solution) {
-          belong_to_a_cluster = true;
-          // Danger Zone for future ref
-          // cluster_id should never be equal to element_solution
-          unsigned int cluster_id = element_solution;
-          elem->set_extra_integer(index, cluster_id);
-          break;
+    neighbor_stack.push(elem);
+    int cluster_id = elem->id();
+
+    while (!neighbor_stack.empty()) {
+      libMesh::Elem *test_elem = neighbor_stack.top();
+      neighbor_stack.pop();
+      if (test_elem->get_extra_integer(index) == not_visited &&
+          BelongToCluster(elem, test_elem, variable_num, system)) {
+
+        test_elem->set_extra_integer(index, cluster_id);
+        for (unsigned int s = 0; s < test_elem->n_sides(); s++) {
+          libMesh::Elem *neighbor_elem = test_elem->neighbor_ptr(s);
+          if (neighbor_elem &&
+              neighbor_elem->get_extra_integer(index) == not_visited) {
```

Comment:
Comment: this test is not necessary, it would be handled by line 96 if we skip it here, but it may have some performance benefit.